### ISSUE TITLE
"Too many positional arguments" for pydantic dataclass constructor

### ISSUE THREAD

**Issue by Ryang20718** [astral-sh/ty#598]
### Summary

```py
from pydantic.dataclasses import dataclass as dataclass_with_validation

@dataclass_with_validation
class BoundingBox3d:
    """Bounding box object.

    Attributes:
        x_size_m (float): Size of bounding box along x-axis.
        y_size_m (float): Size of bounding box along y-axis.
        z_size_m (float): Size of bounding box along z-axis.
    """

    x_size_m: float
    y_size_m: float
    z_size_m: float

print(BoundingBox3d(2.0, 4.0, 6.0))
```

```
WARN ty is pre-release software and not ready for production use. Expect to encounter bugs, missing features, and fatal errors.
error[too-many-positional-arguments]: Too many positional arguments to bound method `__init__`: expected 0, got 3
  --> /tmp/a.py:17:21
   |
15 |     z_size_m: float
16 |
17 | print(BoundingBox3d(2.0, 4.0, 6.0))
   |                     ^^^
   |
info: `too-many-positional-arguments` is enabled by default

Found 1 diagnostic
```

### Version

0.0.1a8

**Comment by carljm** [comment]
Hi, thanks for the report. I'm afraid we'll need a bit more information to do anything with this bug report -- the key question is how the class `BoundingBox3d` is defined.

**Comment by MichaReiser** [comment]
A good indicator that our diagnostic should include more information 😅

**Comment by Ryang20718** [comment]
I'm so sorry, I think the copy - paste repro didn't go through the first time. 

added

**Comment by AlexWaygood** [comment]
`pydantic.dataclasses.dataclass()` is defined [here](https://github.com/pydantic/pydantic/blob/88c75cd5adbfbfea4ff1aceb597c93f238db8dcb/pydantic/dataclasses.py#L29-L319). It uses overloads decorated with `dataclass_transform`, which is quite complicated, but which I would expect to work following https://github.com/astral-sh/ruff/pull/17835. We had a previous issue regarding `pydantic.dataclasses.dataclass` being inferred as `Never`, but that was also closed: https://github.com/astral-sh/ty/issues/307.

**Comment by AlexWaygood** [comment]
@Ryang20718, how exactly are you running ty? I can't reproduce this check locally if I run `uv run --no-project --with=pydantic --with=ty ty check foo.py` (where `foo.py` contains the snippet you pasted above). But I _can_ reproduce if it I run `uv run --no-project --with=ty ty check foo.py`.

Do you definitely have pydantic installed into a virtual environment? If so, where is the virtual environment located? Are you invoking ty directly, or via a project manager such as uv? Are you passing the `--python` flag to ty?

**Comment by Ryang20718** [comment]
We do something of the nature and pass `ty check --extra-search-path /tmp/testing/.venv/lib/python3.12/site-packages <file>` which leads to this error. However, it works if we do 

`ty check --extra-search-path /tmp/testing/.venv/lib/python3.12/site-packages/pydantic <file>`

We do something of the nature for mypy where if we invoke mypy via python (i.e python -m mypy) the PYTHONPATH of our python interpreter propagates and I couldn't quite figure out how to do that in ty 😅. Apologies here. We use bazel so we have to wrangle a lot of tools to work with our custom workspace
